nexus/db-queries/src/db/datastore/ipv4_nat_entry.rs

```diff
@@ -0,0 +1,495 @@
+            time_deleted = now()
+            WHERE time_deleted IS NULL AND version_removed IS NULL AND id = $1 AND version_added = $2
+            ",
+        )
+            .bind::<Uuid, _>(nat_entry.id)
+            .bind::<BigInt, _>(nat_entry.version_added)
+            .execute_async(&*self.pool_connection_authorized(opctx).await?)
+            .await
+            .map_err(|e| public_error_from_diesel(e, ErrorHandler::Server))?;
+
+        if updated_rows == 0 {
+            return Err(Error::ObjectNotFound {
+                type_name: ResourceType::Ipv4NatEntry,
+                lookup_type: LookupType::ByCompositeId(
+                    "id, version_added".to_string(),
+                ),
+            });
+        }
+        Ok(())
+    }
+
+    pub async fn ipv4_nat_cancel_delete(
+        &self,
+        opctx: &OpContext,
+        nat_entry: &Ipv4NatEntry,
+    ) -> DeleteResult {
+        let version_removed =
+            nat_entry.version_removed.ok_or(Error::InvalidRequest {
+                message: "supplied nat entry does not have version_deleted"
+                    .to_string(),
+            })?;
+
+        let updated_rows = sql_query(
+            "
+            UPDATE omicron.public.ipv4_nat_entry
+            SET
+            version_removed = NULL,
+            time_deleted = NULL,
+            version_added = nextval('omicron.public.ipv4_nat_version')
+            WHERE version_removed IS NOT NULL AND id = $1 AND version_removed = $2
+            ",
+        )
+            .bind::<Uuid, _>(nat_entry.id)
+            .bind::<BigInt, _>(version_removed)
+            .execute_async(&*self.pool_connection_authorized(opctx).await?)
+            .await
+            .map_err(|e| public_error_from_diesel(e, ErrorHandler::Server))?;
+
+        if updated_rows == 0 {
+            return Err(Error::InvalidRequest {
+                message: "no matching records".to_string(),
+            });
+        }
+        Ok(())
+    }
+
+    pub async fn ipv4_nat_find_by_id(
+        &self,
+        opctx: &OpContext,
+        id: uuid::Uuid,
+    ) -> LookupResult<Ipv4NatEntry> {
+        use db::schema::ipv4_nat_entry::dsl;
+
+        let result = dsl::ipv4_nat_entry
+            .filter(dsl::id.eq(id))
+            .select(Ipv4NatEntry::as_select())
+            .limit(1)
+            .load_async(&*self.pool_connection_authorized(opctx).await?)
+            .await
+            .map_err(|e| public_error_from_diesel(e, ErrorHandler::Server))?;
+
+        if let Some(nat_entry) = result.first() {
+            Ok(nat_entry.clone())
+        } else {
+            Err(Error::InvalidRequest {
+                message: "no matching records".to_string(),
+            })
+        }
+    }
+
+    pub async fn ipv4_nat_delete_by_external_ip(
+        &self,
+        opctx: &OpContext,
+        external_ip: &ExternalIp,
+    ) -> DeleteResult {
+        let updated_rows = sql_query(
+            "
```

Comment:
See the prior two examples -- this can use the `ipv4_nat_next_version()` function and otherwise avoid defining a full manual SQL query